<launch>
  <arg name="uav_name" default="$(optenv UAV_NAME uav2)"/>

  <arg name="threshold" default="150"/>

  <arg name="device" default="25001879"/>
  <arg name="rate" default="100"/>
  <arg name="serial" default="$(arg device)"/>
  <arg name="camera_name" default="bluefox"/>
  <arg name="camera" default="$(arg camera_name)"/>
  <arg name="frame_id" default="$(arg camera)"/>
  <arg name="calib_url" default="file://$(find mrs_main)/camera_info/bluefox_$(arg uav_name).yaml"/>
  <arg name="fps" default="$(arg rate)"/>
  <arg name="idpf" default="0"/>
  <arg name="aec" default="false"/>
  <arg name="expose_us" default="1500"/>
  <arg name="agc" default="false"/>
  <arg name="gain_db" default="0.0"/>
  <arg name="cbm" default="0"/>
  <arg name="ctm" default="1"/>
  <arg name="dcfm" default="0"/>
  <arg name="hdr" default="false"/>
  <arg name="wbp" default="-1"/>
  <arg name="request" default="0"/>
  <arg name="mm" default="0"/>
  <arg name="jpeg_quality" default="90"/>

  <arg name="processRate" default="10"/>

  <arg name="accumulatorLength"         default="23"/>
  <arg name="pitchSteps"         default="16"/>
  <arg name="yawSteps"         default="8"/>
  <arg name="maxPixelShift"         default="2"/>

  <arg name="publishVisualization"         default="true"/>
  <arg name="visualizationRate"         default="5"/>

  <arg name="filterDistLength"         default="10"/>
  <arg name="filterOrientationLength"         default="10"/>

  <!-- Node Settings -->
  <arg name="output" default="screen"/>
  <arg name="proc" default="false"/>
  <arg name="view" default="false"/>
  <arg name="calib" default="false"/>

  <arg name="DEBUG" default="false"/>
  <arg name="VisDEBUG" default="false"/>
  <arg name="gui" default="false"/>

  <arg name="trajectoryControl" default="true"/>

  <arg name="followDistance" default="5"/>
  <arg name="trajCoeff" default="0.8"/>
  <arg name="yawCoeff" default="0.1"/>
  <arg name="tailingCoeff" default="1.0"/>

  <arg name="SamplePointSize" default="8"/>
  <arg name="StepSize" default="10"/>
  <arg name="publish" default="true"/>
  <arg name="useOdom" default="false"/>
  <arg name="cemeraRotated" default="false"/>
  <arg name="FromBag" default="false"/>
  <arg name="FromCamera" default="true"/>
  <arg name="FromVideo" default="false"/>
  <arg name="VideoNumber" default="9"/>
  <arg name="cellSize" default="64"/>
  <arg name="cellOverlay" default="16"/>
  <arg name="surroundRadius" default="3"/>
  <arg name="camNum" default="0"/>
  <arg name="useFAST" default="true"/>
  <arg name="Delay" default="100"/>
  <arg name="TimeScale" default="1.0"/>


  <node
    name="uvcam_tf"
    pkg="tf"
    type="static_transform_publisher"
    args="0.108 -0.06 0.185 -1.57079632679 0.00000000000 -1.57079632679 fcu_$(arg uav_name) uvcam_$(arg uav_name) 100"/>

  <!-- <origin xyz="0.108 0 0.085" rpy="1.57079632679 4.71238898038 1.57079632679" /> -->
  <!--Usage: static_transform_publisher x y z yaw pitch roll frame_id child_frame_id  period(milliseconds)-->
<!-- OR -->
  <group ns="$(arg uav_name)">

    <node name="uvdd_detect" pkg="uvdd" type="uv_detector_node" output="screen"> 
      <!-- <node name="uvdd_detect" pkg="uvdd" type="uv_detector_node" output="screen" launch-prefix="urxvt -e gdb -q -x /home/viktor/gdb.cmds -/-args"> -->
        <param name="uav_name" type = "string" value="$(arg uav_name)"/>
        <param name="DEBUG"		type="bool" value="$(arg DEBUG)"/> 
        <param name="gui"             type="bool" value="false"/>
        <param name="Flip"             type="bool" value="false"/>
        <param name="useFAST"             type="bool" value="$(arg useFAST)"/>
        <!-- <param name="MaskPath"             type="string" value="$(arg MaskPath)"/> -->
        <param name="FromVideo"         type="bool" value="$(arg FromVideo)"/>
        <param name="FromBag"         type="bool" value="$(arg FromBag)"/>
        <param name="FromCamera"         type="bool" value="$(arg FromCamera)"/>
        <param name="camTopic"         type="string" value="camera"/>
        <param name="VideoNumber"     type="int" value="$(arg VideoNumber)"/>
        <param name="SamplePointSize" type="int" value="$(arg SamplePointSize)"/>
        <param name="StepSize"	type="int" value="$(arg StepSize)"/>
        <param name="cellSize"	type="int" value="$(arg cellSize)"/>
        <param name="cellOverlay"	type="int" value="$(arg cellOverlay)"/>
        <param name="surroundRadius"	type="int" value="$(arg surroundRadius)"/>
        <param name="publish"         type="bool" value="$(arg publish)"/>
        <param name="cameraRotated"    type="bool" value="false"/>
        <param name="useOdom"           type="bool" value="false"/>
        <param name="camNum"           type="int" value="$(arg camNum)"/>
        <param name="justReport" type="bool" value="true"/>
        <param name="threshold" type="int" value="$(arg threshold)"/>
        <param name="CameraImageCompressed" type="bool" value="false"/>

        <remap from="odometry" to="/$(arg uav_name)/mrs_odometry/new_odom"/>
        <remap from="camera" to="/$(arg uav_name)/bluefox_uv/image_raw"/>
        <remap from="~imu" to="mavros/imu/data"/>
        <remap from="~pointsSeen" to="/$(arg uav_name)/uvPointsSeen"/>
        <rosparam command="load" file="$(find uvdd)/yaml/bluefox_fisheye.yaml" />
      </node>
      <node name="blink_follow" pkg="blink_processor" type="uv_blink_follower_node" output="screen"> 
        <param name="uav_name" type = "string" value="$(arg uav_name)"/>

        <param name="followDistance" type="double" value="$(arg followDistance)"/>
        <param name="trajCoeff" type="double" value="$(arg trajCoeff)"/>
        <param name="yawCoeff" type="double" value="$(arg yawCoeff)"/>
        <param name="tailingCoeff" type="double" value="$(arg tailingCoeff)"/>

        <param name="filterDistLength" type="double" value="$(arg filterDistLength)"/>
        <param name="filterOrientationLength" type="double" value="$(arg filterOrientationLength)"/>

        <param name="trajectoryControl"             type="bool" value="$(arg trajectoryControl)"/>

        <param name="DEBUG"		type="bool" value="$(arg DEBUG)"/> 
        <param name="gui"             type="bool" value="$(arg gui)"/>
        <param name="publish"         type="bool" value="$(arg publish)"/>
        <param name="useOdom"           type="bool" value="false"/>
        <param name="justReport" type="bool" value="true"/>


        <remap from="odometry" to="/$(arg uav_name)/mrs_odometry/new_odom"/>
        <remap from="diagnostics" to="/$(arg uav_name)/trackers_manager/mpc_tracker/diagnostics"/>
        <remap from="~imu" to="mavros/imu/data"/>
        <remap from="blinkersSeen" to="/$(arg uav_name)/blinkersSeen"/>
        <rosparam command="load" file="$(find uvdd)/yaml/bluefox_fisheye.yaml" />
      </node>
      <node name="blink_processor" pkg="blink_processor" type="blink_processor_node" output="screen"  > 
        <!-- <node name="blink_processor" pkg="blink_processor" type="blink_processor_node" output="screen"  launch-prefix="urxvt -e gdb -q -x /home/viktor/gdb.cmds -\-args"> --> 
          <param name="DEBUG"		type="bool" value="false"/> 
          <param name="VisDEBUG"		type="bool" value="$(arg VisDEBUG)"/> 
          <param name="GUI"             type="bool" value="$(arg gui)"/>
          <param name="CameraImageCompressed"    type="bool" value="false"/>
          <param name="InvertedPoints"    type="bool" value="false"/>
          <param name="frequencyCount" type="int" value="3"/>
          <param name="frequency1" type="int" value="6"/>
          <param name="frequency2" type="int" value="10"/>
          <param name="frequency3" type="int" value="30"/>

          <param name="processRate"         type="int" value="$(arg processRate)"/>

          <param name="accumulatorLength"         type="int" value="$(arg accumulatorLength)"/>
          <param name="pitchSteps"         type="int" value="$(arg pitchSteps)"/>
          <param name="yawSteps"         type="int" value="$(arg yawSteps)"/>
          <param name="maxPixelShift"         type="int" value="$(arg maxPixelShift)"/>
          
          <param name="publishVisualization"         type="bool" value="$(arg publishVisualization)"/>
          <param name="visualizationRate"         type="int" value="$(arg visualizationRate)"/>

          <remap from="camera" to="/$(arg uav_name)/bluefox_uv/image_raw"/>
          <remap from="visualization" to="/$(arg uav_name)/blink_visualization"/>
          <remap from="pointsSeen" to="/$(arg uav_name)/uvPointsSeen"/>
          <remap from="~blinkersSeen" to="/$(arg uav_name)/blinkersSeen"/>

          <rosparam command="load" file="$(find blink_processor)/yaml/bluefox_fisheye.yaml" />

        </node>
        <node pkg="bluefox2" type="bluefox2_single_node" name="bluefox_uv" output="screen">
          <param name="identifier" type="string" value="$(arg serial)"/>
          <param name="frame_id" type="string" value="$(arg frame_id)"/>
          <param name="camera_name" type="string" value="$(arg camera_name)"/>
          <param name="calib_url" type="string" value="$(arg calib_url)"/>
          <param name="fps" type="double" value="$(arg fps)"/>
          <param name="idpf" type="int" value="$(arg idpf)"/>
          <param name="aec" type="bool" value="$(arg aec)"/>
          <param name="expose_us" type="int" value="$(arg expose_us)"/>
          <param name="agc" type="bool" value="$(arg agc)"/>
          <param name="gain_db" type="double" value="$(arg gain_db)"/>
          <param name="cbm" type="int" value="$(arg cbm)"/>
          <param name="ctm" type="int" value="$(arg ctm)"/>
          <param name="dcfm" type="int" value="$(arg dcfm)"/>
          <param name="hdr" type="bool" value="$(arg hdr)"/>
          <param name="wbp" type="int" value="$(arg wbp)"/>
          <param name="request" type="int" value="$(arg request)"/>
          <param name="mm" type="int" value="$(arg mm)"/>

          <!-- Params for theora quality, images being compressed into "stream" -->
          <param name="image_raw/compressed/jpeg_quality" type="int" value="$(arg jpeg_quality)"/>
          <param name="image_raw/theora/keyframe_frequency" value="60" />
          <param name="image_raw/theora/target_bitrate" value="50000" />
          <param name="image_raw/theora/quality" value="8" />
          <param name="image_raw/theora/optimize_for" value="0" />
        </node>
      </group>

      <!-- </group> -->



  </launch>
